llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -886,11 +903,29 @@ bool RISCVVLOptimizer::checkUsers(const MachineOperand *&CommonVL,
 
     // Instructions like reductions may use a vector register as a scalar
     // register. In this case, we should treat it like a scalar register which
-    // does not impact the decision on whether to optimize VL.
-    // TODO: Treat it like a scalar register instead of bailing out.
+    // does not impact the decision on whether to optimize VL. But if there is
+    // another user of MI and it has VL=0, we need to be sure not to reduce the
+    // VL of MI to zero when the VLOp of UserOp is may be non-zero.
     if (isVectorOpUsedAsScalarOp(UserOp)) {
-      CanReduceVL = false;
-      break;
+      [[maybe_unused]] Register R = UserOp.getReg();
+      [[maybe_unused]] const TargetRegisterClass *RC = MRI->getRegClass(R);
+      assert(RISCV::VRRegClass.hasSubClassEq(RC) &&
+             "Expect LMUL 1 register class for vector as scalar operands!");
+      LLVM_DEBUG(dbgs() << "    Used this operand as a scalar operand\n");
+      const MCInstrDesc &Desc = UserMI.getDesc();
+      unsigned VLOpNum = RISCVII::getVLOpNum(Desc);
+      const MachineOperand &VLOp = UserMI.getOperand(VLOpNum);
+      if (VLOp.isReg() || (VLOp.isImm() && VLOp.getImm() != 0)) {
```

Comment:
If it's a register, you can't tell whether it is dynamically 0 or not.

Reply:
I agree. I think the code is handling it correctly.

This is what happens:

```
If VLOp.isReg
  if !CommonVL
    Set CommonVL as VLOP. If another UserMI has VL=0, it will be incompatible with VLOp. If UserMI has VL!=0 then it must match this VLOp and we end up optimizing to VL=VLOP.
 if !CommonVL.isIdenticalTo
   Do not reduce the VL
 if CommonVL isIdenticalTo
   No action is needed since we will use VL=VLOP.
```